phonebook-api/src/main/java/com/telran/phonebookapi/controller/ContactController.java

```diff
@@ -1,67 +1,91 @@
 import javax.validation.Valid;
 import java.util.List;
+import java.util.stream.Collectors;
 
 @RestController
-@CrossOrigin
 @RequestMapping("/api/contact")
 public class ContactController {
 
+    static final String INVALID_ACCESS = "Error! You have no permission";
+
+    UserService userService;
     ContactService contactService;
+    ContactMapper contactMapper;
 
-    public ContactController(ContactService contactService) {
+    public ContactController(UserService userService, ContactService contactService, ContactMapper contactMapper) {
+        this.userService = userService;
         this.contactService = contactService;
+        this.contactMapper = contactMapper;
     }
 
     @PostMapping("")
-    public void addContact(@Valid @RequestBody ContactDto contactDto) {
-        contactService.add(contactDto);
+    public void addContact(Authentication auth, @Valid @RequestBody ContactDto contactDto) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        contactService.add(contactDto.firstName, contactDto.lastName, contactDto.description, userDetails.getUsername());
     }
 
     @GetMapping("/{id}")
-    public ContactDto getById(@PathVariable int id) {
-        return contactService.getById(id);
-    }
-
-    @GetMapping("/{id}/extended")
-    public ContactDto getByIdFullDetails(@PathVariable int id) {
-        return contactService.getByIdFullDetails(id);
+    public ContactDto getById(Authentication auth, @PathVariable int id) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        Contact contact = contactService.getById(id);
+        if (!contact.getUser().getEmail().equals(userDetails.getUsername())) {
+            throw new EntityNotFoundException(INVALID_ACCESS);
+        }
+        return contactMapper.mapContactToDtoFull(contactService.getById(id));
     }
 
     @PutMapping("")
-    public void editContact(@Valid @RequestBody ContactDto contactDto) {
-        contactService.editAllFields(contactDto);
+    public void editContact(Authentication auth, @Valid @RequestBody ContactDto contactDto) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        Contact contact = contactService.getById(contactDto.id);
+        if (!contact.getUser().getEmail().equals(userDetails.getUsername())) {
+            throw new EntityNotFoundException(INVALID_ACCESS);
+        }
+        contactService.editContact(contactDto.firstName, contactDto.lastName, contactDto.description, contactDto.id);
     }
 
     @DeleteMapping("/{id}")
-    public void removeById(@PathVariable int id) {
+    public void removeById(Authentication auth, @PathVariable int id) {
+        UserDetails userDetails = (UserDetails) auth.getPrincipal();
+        Contact contact = contactService.getById(id);
+        if (!contact.getUser().getEmail().equals(userDetails.getUsername())) {
+            throw new EntityNotFoundException(INVALID_ACCESS);
+        }
         contactService.removeById(id);
```

Comment:
make sure that along with the contact all its parts (like phones, addresses ect.) are removed

Reply:
done, all its parts were removed